File: src/stress_stream/consumer.py
import asyncio
import struct
from datetime import datetime

import typer
from prometheus_client import Counter, Histogram
from streamstore import S2
from streamstore.schemas import SequencedRecord

from stress_stream._constants import HISTOGRAM_BUCKETS, METRICS_NAMESPACE, S2_AUTH_TOKEN
from stress_stream._utils import metrics_server

e2e_latency_histogram = Histogram(
    name="e2e_latency_seconds",
    documentation="End-to-end latency in seconds. This measures the time taken from when a record is produced "
    "(by the producer) to when it is read by the consumer.",
    namespace=METRICS_NAMESPACE,
    buckets=HISTOGRAM_BUCKETS,
)
read_records_counter = Counter(
    name="read_records_total",
    documentation="Total number of records successfully read.",
    namespace=METRICS_NAMESPACE,
)
read_bytes_counter = Counter(
    name="read_bytes_total",
    documentation="Total number of bytes successfully read.",
    namespace=METRICS_NAMESPACE,
)
reads_counter = Counter(
    name="reads_total",
    documentation="Total number of reads in read sessions.",
    namespace=METRICS_NAMESPACE,
)
read_failures_counter = Counter(
    name="read_failures_total",
    documentation="Total number of failed reads in read sessions.",
    namespace=METRICS_NAMESPACE,
)


def observe_metrics(record: SequencedRecord):
    name, val = record.headers[0]
    read_bytes_counter.inc(len(record.body) + len(name) + len(val))
    created_ts = struct.unpack("d", val)[0]
    e2e_latency = (datetime.now() - datetime.fromtimestamp(created_ts)).total_seconds()
    e2e_latency_histogram.observe(e2e_latency)


async def consumer(basin_name: str, stream_name: str):
    async with S2(auth_token=S2_AUTH_TOKEN) as s2:
        stream = s2[basin_name][stream_name]
        start_seq_num = await stream.check_tail()
        while True:
            try:
                reads_counter.inc()
                async for output in stream.read_session(start_seq_num):
                    reads_counter.inc()
                    match output:
                        case list(records):
                            if len(records) > 0:
                                start_seq_num = records[-1].seq_num + 1
                            read_records_counter.inc(len(records))
                            for record in records:
                                observe_metrics(record)
                        case unexpected:
                            raise RuntimeError(
                                "Unexpected output when reading from stream. "
                                f"Expected a batch of records, but got {unexpected}"
                            )
            except Exception:
                read_failures_counter.inc()


async def consumer_and_metrics_server(basin_name: str, stream_name: str):
    async with asyncio.TaskGroup() as tg:
        tg.create_task(consumer(basin_name, stream_name))
        tg.create_task(metrics_server())


def main(
    basin: str = typer.Option(..., help="Name of the basin."),
    stream: str = typer.Option(..., help="Name of the stream."),
):
    asyncio.run(consumer_and_metrics_server(basin, stream))


if __name__ == "__main__":
    typer.run(main)

File: rust/src/bin/consumer.rs
use clap::Parser;
use eyre::eyre;
use futures::StreamExt;
use s2::{
    types::{ReadOutput, ReadSessionRequest},
    Client, ClientConfig, StreamClient,
};
use std::env;
use stress_stream::{
    current_timestamp, init_rustls, init_tracing, metrics_server, BASIN_LABEL, METRICS_NAMESPACE,
    STREAM_LABEL, TIMESTAMP_HEADER_NUM_BYTES,
};
use tracing::{error, info};

fn read_records_counter(basin: String, stream: String) -> metrics::Counter {
    metrics::counter!(
        format!("{METRICS_NAMESPACE}_read_records_total"),
        BASIN_LABEL => basin,
        STREAM_LABEL => stream
    )
}

fn read_bytes_counter(basin: String, stream: String) -> metrics::Counter {
    metrics::counter!(
        format!("{METRICS_NAMESPACE}_read_bytes_total"),
        BASIN_LABEL => basin,
        STREAM_LABEL => stream
    )
}

fn e2e_latency_histogram(basin: String, stream: String) -> metrics::Histogram {
    metrics::histogram!(
        format!("{METRICS_NAMESPACE}_e2e_latency_seconds"),
        BASIN_LABEL => basin,
        STREAM_LABEL => stream,
    )
}

fn read_sessions_counter(basin: String, stream: String) -> metrics::Counter {
    metrics::counter!(
        format!("{METRICS_NAMESPACE}_read_sessions_total"),
        BASIN_LABEL => basin,
        STREAM_LABEL => stream
    )
}

fn read_session_failures_counter(basin: String, stream: String) -> metrics::Counter {
    metrics::counter!(
        format!("{METRICS_NAMESPACE}_read_session_failures_total"),
        BASIN_LABEL => basin,
        STREAM_LABEL => stream,
    )
}

async fn consume_records(
    basin: String,
    stream: String,
    client: StreamClient,
    start_seq_num: &mut u64,
) -> eyre::Result<()> {
    read_sessions_counter(basin.clone(), stream.clone()).increment(1);
    let mut read_stream = client
        .read_session(ReadSessionRequest::new(*start_seq_num))
        .await?;
    while let Some(output) = read_stream.next().await {
        let output = output?;
        match output {
            ReadOutput::Batch(batch) => {
                let mut batch_num_bytes = 0;
                for record in &batch.records {
                    if record.headers.len() != 1 {
                        return Err(eyre!("Unexpected number of headers"));
                    }
                    let header = &record.headers[0];
                    let created_ts = f64::from_be_bytes(header.value.as_ref().try_into()?);
                    e2e_latency_histogram(basin.clone(), stream.clone())
                        .record((current_timestamp()? - created_ts).max(0.0));
                    batch_num_bytes += record.body.len() as u64 + TIMESTAMP_HEADER_NUM_BYTES;
                }
                read_bytes_counter(basin.clone(), stream.clone()).increment(batch_num_bytes);
                read_records_counter(basin.clone(), stream.clone())
                    .increment(batch.records.len() as u64);
                if let Some(record) = batch.records.last() {
                    *start_seq_num = record.seq_num + 1;
                }
            }
            _ => return Err(eyre!("Unexpected output when reading from stream")),
        }
    }
    Ok(())
}

async fn run_consumer(auth_token: String, basin: String, stream: String) -> eyre::Result<()> {
    let config = ClientConfig::new(auth_token);
    let client = Client::new(config)
        .basin_client(basin.clone().try_into()?)
        .stream_client(stream.clone());
    let mut start_seq_num = client.check_tail().await?;
    info!("starting consumption");
    loop {
        let result = consume_records(
            basin.clone(),
            stream.clone(),
            client.clone(),
            &mut start_seq_num,
        )
        .await;
        if result.is_err() {
            error!(err=?result.err().unwrap(), "read session failed");
            read_session_failures_counter(basin.clone(), stream.clone()).increment(1);
        }
    }
}

#[derive(Parser, Debug)]
struct Args {
    #[arg(long, help = "Name of the basin")]
    basin: String,
    #[arg(long, help = "Name of the stream")]
    stream: String,
}

#[tokio::main]
async fn main() -> eyre::Result<()> {
    init_tracing();
    init_rustls();

    let args = Args::parse();
    let auth_token = env::var("S2_AUTH_TOKEN").expect("S2_AUTH_TOKEN env var should be set");
    tokio::task::spawn(metrics_server());
    run_consumer(auth_token, args.basin, args.stream).await?;
    Ok(())
}
